driver_test.go

```diff
@@ -338,3 +309,260 @@ func TestIsDdl(t *testing.T) {
+			wantError: true,
+		},
+		{
+			name: "insert lower case dml",
+			input: `
+			insert into TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (4, "four", CAST("four" as bytes), 42, 42, true ) `,
+			query: `SELECT * FROM TestExecContextDml WHERE key = 4`,
+			wantRows: []testExecContextDmlRow{
+				{key: 4, testString: "four", testBytes: []byte("four"), testInt: 42, testFloat: 42, testBool: true},
+			},
+		},
+		{
+			name: "insert lower case table name",
+			input: `
+			INSERT INTO testexeccontextdml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (5, "five", CAST("five" as bytes), 42, 42, true ) `,
+			query: `SELECT * FROM TestExecContextDml WHERE key = 5`,
+			wantRows: []testExecContextDmlRow{
+				{key: 5, testString: "five", testBytes: []byte("five"), testInt: 42, testFloat: 42, testBool: true},
+			},
+		},
+		{
+			name: "wrong types",
+			input: `
+			INSERT INSERT INTO TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (102, 56, 56, 42, hello, "i should be a bool" ) `,
+			wantError: true,
+		},
+		{
+			name: "insert primary key duplicate",
+			input: `
+			INSERT INSERT INTO TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (1, "one", CAST("one" as bytes), 42, 42, true ) `,
+			wantError: true,
+		},
+		{
+			name: "insert null into primary key",
+			input: `
+			INSERT INSERT INTO TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (null, "one", CAST("one" as bytes), 42, 42, true ) `,
+			wantError: true,
+		},
+		{
+			name: "insert too many values",
+			input: `
+			INSERT INSERT INTO TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (null, "one", CAST("one" as bytes), 42, 42, true, 42 ) `,
+			wantError: true,
+		},
+		{
+			name: "insert too few values",
+			input: `
+			INSERT INSERT INTO TestExecContextDml 
+				(key, testString, testBytes, testInt, testFloat, testBool)
+				VALUES (null, "one", CAST("one" as bytes), 42 ) `,
+			wantError: true,
+		},
+		{
+			name:     "delete single tuple",
+			input:    `DELETE FROM TestExecContextDml WHERE key = 1`,
+			query:    `SELECT * FROM TestExecContextDml WHERE key = 1`,
+			wantRows: []testExecContextDmlRow{},
+		},
+		{
+			name: "delete multiple tuples with subquery",
+			input: `
+			DELETE FROM TestExecContextDml WHERE key 
+				IN (SELECT key FROM TestExecContextDml WHERE key != 4)`,
+			query:    `SELECT * FROM TestExecContextDml WHERE key = 2 OR key = 3`,
+			wantRows: []testExecContextDmlRow{},
+		},
+		{
+			name: "delete all remaining tuples with NOT IN",
+			input: `
+			DELETE FROM TestExecContextDml WHERE key 
+				NOT IN (101)`,
+			query:    `SELECT * FROM TestExecContextDml`,
+			wantRows: []testExecContextDmlRow{},
+		},
+		{
+			name: "refil table for update tests",
```

Comment:
It seems like we are adding a dependency here, from this test to the next ones. IMHO tests should be self contained, so that the order in which they are executed do not influence one another. Furthermore, we should be able to run them independently from one another. So, each test case should:

- Start with an empty table
- Insert the necessary data for the test
- Run the test query
- Assert on the results

Could we perhaps change the data structure to support such plan?

Reply:
Sure thing, I'll add a set up field in the test data struct & clear the table at the end of the run tests loop, let me know if that's weird

Reply:
It makes sense to me, thanks!

Reply:
changed :)